daffodil-runtime1/src/main/scala/org/apache/daffodil/processors/VariableMap1.scala

```diff
@@ -129,54 +156,66 @@ final class VariableBox(initialVMap: VariableMap) {
  * The DPath implementation must be made to implement the
  * no-set-after-default-value-has-been-read behavior. This requires that reading the variables causes a state transition.
  */
-class VariableMap private(vTable: Map[GlobalQName, List[List[Variable]]])
+class VariableMap private(vTable: Map[GlobalQName, MStackOf[Variable]])
   extends Serializable {
 
   def this(topLevelVRDs: Seq[VariableRuntimeData] = Nil) =
-    this(topLevelVRDs.map {
+    this(Map(topLevelVRDs.map {
       vrd =>
         val variab = vrd.newVariableInstance
-        (vrd.globalQName, List(List(variab)))
-    }.toMap)
+        val stack = new MStackOf[Variable]
+        stack.push(variab)
+        (vrd.globalQName, stack)
+    }: _*))
 
   override def toString(): String = {
     "VariableMap(" + vTable.mkString(" | ") + ")"
   }
 
+  def copy(): VariableMap = {
+    val table = Map[GlobalQName, MStackOf[Variable]]()
+    vTable.foreach { case (k: GlobalQName, s: MStackOf[Variable]) => {
+      val newStack= new MStackOf[Variable]()
+      newStack.push(s.top.copy())
+      table(k) = newStack
+    }}
+
+    new VariableMap(table)
+  }
+
   def find(qName: GlobalQName): Option[Variable] = {
     val optVrd = getVariableRuntimeData(qName)
-    val optLists = optVrd.flatMap { vrd => vTable.get(vrd.globalQName) }
-    val variab = optLists.flatMap { lists => lists.flatMap {
-      _.toStream
-    }.headOption
+    val optStack = optVrd.flatMap { vrd => vTable.get(vrd.globalQName) }
+    val variab = {
+      if (optStack.isDefined)
+        Some(optStack.get.top)
+      else
+        None
```

Comment:
Can this just instead just do something like:
```scala
val optStack = vTable.get(qName)
val variab = ...
```
If this uses getVariableRuntimeData it's going to do two looks up. Once in getVariableRuntimeData to find the right stack and get the rd of the the thing on the top, then another to find that stack that getVariableRuntimeData already found.

And then getVariableRuntimeData can just be
```
val optVariable = find(qName)
if (optVariable.isDefined) Some(optVariable.get.rd) else None
```